File: Platform/Workspace/Typescript/framework/data/Exists.ts
import { PropertyType, ObjectType } from '../meta';
import { ParameterizablePredicate, ParameterizablePredicateArgs } from './ParameterizablePredicate';

export interface ExistArgs extends ParameterizablePredicateArgs, Pick<Exists, 'propertyType'> {}

export class Exists extends ParameterizablePredicate {
  propertyType: PropertyType;

  constructor(propertyType: PropertyType);
  constructor(args: ExistArgs);
  constructor(args: ExistArgs | PropertyType) {
    super();

    if ((args as PropertyType).objectType) {
      this.propertyType = args as PropertyType;
    } else {
      Object.assign(this, args);
      this.propertyType = (args as ExistArgs).propertyType;
    }
  }

  get objectType(): ObjectType {
    return this.propertyType.objectType;
  }

  public toJSON(): any {
    return {
      kind: 'Exists',
      dependencies: this.dependencies,
      propertytype: this.propertyType.id,
      parameter: this.parameter,
    };
  }
}

File: Platform/Workspace/CSharp/Allors.Workspace/Data/Exists.cs
// <copyright file="Exists.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Data
{
    using Allors.Protocol.Data;
    using Allors.Workspace.Meta;

    public class Exists : IPropertyPredicate
    {
        public string[] Dependencies { get; set; }

        public Exists(IPropertyType propertyType = null) => this.PropertyType = propertyType;

        public string Parameter { get; set; }

        public IPropertyType PropertyType { get; set; }

        public Predicate ToJson() =>
            new Predicate
            {
                kind = PredicateKind.Exists,
                dependencies = this.Dependencies,
                propertyType = this.PropertyType?.Id,
                parameter = this.Parameter,
            };
    }
}
